File: api/app/models/product.py
from libdev.lang import to_url

from models import Base, Attribute


def default_url(instance):
    """Generate default URL slug based on title and id"""
    url = to_url(instance.title) or ""
    if url:
        url += "-"
    return url + f"{instance.id}"


class Product(Base):
    _name = "products"
    _search_fields = {"title", "description", "category"}

    title = Attribute(types=str)
    description = Attribute(types=str, default="")
    images = Attribute(types=list, default=list)
    price = Attribute(types=float, default=0.0)
    discount_type = Attribute(types=str)
    discount_value = Attribute(types=float)
    currency = Attribute(types=str, default="$")
    rating = Attribute(types=float)
    rating_count = Attribute(types=int)
    category = Attribute(types=str)
    in_stock = Attribute(types=bool, default=True)
    is_new = Attribute(types=bool, default=False)
    is_featured = Attribute(types=bool, default=False)
    features = Attribute(types=list)
    options = Attribute(types=list)
    url = Attribute(types=str, default=default_url)
    status = Attribute(types=int, default=1)
    token = Attribute(types=str)

File: web/src/entities/product/model/product.ts
export interface Product {
  id: number;
  title: string;
  description?: string;
  images: string[];
  priceFrom: number;
  finalPriceFrom?: number;
  price?: number;
  options: ProductOption[];
  features?: ProductFeature[];
  currency?: string;
  rating?: number;
  ratingCount?: number;
  category?: string;
  inStock?: boolean;
  isNew?: boolean;
  isFeatured?: boolean;
  created?: number;
  updated?: number;
  status?: number;
  url?: string;
}

export type ProductFeatureValueType = 'string' | 'number' | 'boolean';

export interface ProductFeature {
  key: string;
  value: string | number | boolean;
  valueType: ProductFeatureValueType;
}

export interface ProductOption {
  name: string;
  price: number;
  finalPrice?: number;
  discountType?: 'percent' | 'fixed';
  discountValue?: number;
  images: string[];
  rating?: number;
  ratingCount?: number;
  inStock?: boolean;
  attributes?: ProductFeature[];
  features?: ProductFeature[];
}

export interface ProductsGetRequest {
  id?: number | number[];
  search?: string;
  category?: string;
  limit?: number;
  offset?: number;
}

export interface ProductsGetResponse {
  products: Product[];
  count?: number;
}

export interface ProductSaveRequest {
  id?: number;
  title: string;
  description?: string;
  images?: string[];
  features?: ProductFeature[];
  options: Array<{
    name: string;
    price: number;
    discountType?: 'percent' | 'fixed';
    discountValue?: number;
    images?: string[];
    rating?: number;
    ratingCount?: number;
    inStock?: boolean;
    attributes?: ProductFeature[];
    features?: ProductFeature[];
  }>;
  currency?: string;
  category?: string;
  isNew?: boolean;
  isFeatured?: boolean;
  status?: number;
}
